### DISCUSSION TITLE
@BeforeEach를 사용한 테스트의 장단점은 무엇일까요?

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#101] [Test]
.

**Comment by alstn113** [comment]
`@BeforeEach` 어노테이션은 Junit 프레임워크에서 사용되는 어노테이션 중 하나로, 각각의 테스트 메서드가 실행되기 전에 매번 실행되는 메서드를 지정할 때 사용됩니다.

`@BeforeEach`를 사용한 테스트의 장단점은 다음과 같습니다.

장점
1. 각각의 테스트 메서드에서 반복되는 설정 코드를 중복으로 작성할 필요가 없어집니다. 대신 `@BeforeEach` 어노테이션이 지정된 메서드에서 설정 코드를 한 번만 작성하면 됩니다.
2. 각 테스트가 독립적으로 실행되도록 보장할 수 있습니다. 이는 테스트 간의 의존성을 최소화하여 테스트의 신뢰성을 향상시킵니다.
3. 테스트 메서드 자체에는 테스트 목적에만 집중할 수 있습니다. 테스트 메서드에서 테스트의 설정이나 준비 작업을 직접 작성하는 것보다 가독성이 향상됩니다.

단점
1. 어떤 상황일 경우에 이 기대값이 나오는지를 한 눈에 알 수가 없다. 그래서 setup()메서드를 확인해봐야 하다.
2. 해당 테스트 클래스에서 모든 테스트가 결합되어 버린다. `@BeforeEach` 하나만 변경해도 모든 메서드가 영향을 받는다.

**Comment by GIVEN53** [comment]
### 장점
1. 반복 코드 제거
    - 각각의 테스트 메서드에서 반복되는 초기화 코드를 제거할 수 있어 코드의 중복을 줄이고 유지보수성을 향상시킨다.
2. 테스트 코드 가독성 향상
    - 초기화 코드가 각각의 테스트 메서드에 포함되지 않으므로 테스트 메서드 자체의 목적을 더 명확하게 이해할 수 있다.
3. 테스트 유연성 증가
    - 각각의 테스트 메서드가 독립적으로 실행될 수 있으므로, 테스트 간의 의존성을 최소화하고 테스트의 순서를 변경해도 영향을 받지 않는다.
4. 테스트 유지보수 용이성
    - 초기화 코드가 하나의 메서드에 집중되기 때문에 초기화 로직의 변경이 필요할 때 해당 메서드만 수정하면 된다.

### 단점
1. 테스트 간 의존성 증가
    - `@BeforeEach`를 남용하면 각각의 테스트가 서로 의존적일 수 있다. 테스트 간의 결합도를 증가시켜 테스트의 격리성을 감소시킬 수 있다.
2. 테스트 실행 시간 증가
    - 각각의 테스트 메서드가 실행되기 전에 초기화 코드가 실행되기 때문에 초기화 작업이 복잡하거나 많은 리소스를 필요로 하는 경우 테스트의 실행 시간이 증가할 수 있다.
3. 초기화 메서드 오용
    - 모든 테스트 메서드가 해당 초기화 메서드를 실행하기 때문에 테스트 간의 관계가 불필요하게 복잡해질 수 있다.
4. 테스트 의도 모호성
    - 초기화 로직이 테스트 메서드와 분리되어 테스트의 의도가 명확하지 않을 수 있다.

**Comment by le2sky** [comment]
`@BeforeEach`를 사용하면  각 테스트 케이스가 실행되기 전에 설정 코드가 동작합니다.
이때, 테스트 케이스들은 설정 코드 내부의 모든 설정을 공유하게 되는데요.
이는 양날의 검이 될 수 있습니다.
모든 테스트 케이스가 필요한 사전 작업이 복잡한 경우 중복 코드를 제거하여, 테스트를 단순화할 수 있습니다.
하지만, 특정 테스트 케이스에 대한 설정을 위해서 설정 코드를 수정하는 경우에 해당 설정 코드에 의존하는 
다른 테스트 케이스에게 영향을 줄 수 있습니다.
또한, 사전 작업이 테스트 메서드로부터 멀리 떨어진 곳에 존재하기 때문에, 테스트가 무엇을 검증하는지 이해하기 어려울 수 있습니다.

**Comment by GIVEN53** [comment]
중복 코드 제거, 가독성 향상
의존성 증가, 의도 모호성
